KEP-3926: unsafe deletion of corrupt objects

tkashem opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
A new feature that allows unsafe deletion of corrupt resources has been added, it is disabled by default, 
and it can be enabled by setting the option `--feature-gates=AllowUnsafeMalformedObjectDeletion=true`. 
It comes with an API change, a new delete option `ignoreStoreReadErrorWithClusterBreakingPotential` has
been introduced, it is not set by default, this maintains backward compatibility. 
In order to perform an unsafe deletion of a corrupt resource, the user must enable the option for the delete
request. A resource is considered corrupt if it can not be successfully retrieved from the storage due to
a) transformation error e.g. decryption failure, or b) the object failed to decode. Normal deletion flow is
attempted first, and if it fails with a corrupt resource error then it triggers unsafe delete.
In addition, when this feature is enabled, the 'details' field of 'Status' from the LIST response 
includes information that identifies the corrupt object(s).
NOTE: unsafe deletion ignores finalizer constraints, and skips precondition checks.
WARNING: this may break the workload associated with the resource being unsafe-deleted, if it relies on
the normal deletion flow, so cluster breaking consequences apply.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

KEP: https://github.com/kubernetes/enhancements/pull/3927

```docs

```

tkashem committed:
implement unsafe deletion, and wire it

- implement unsafe deletion, and wire it
- aggregate corrupt object error(s) from the storage LIST operation
- extend storage error:
a) add a new type ErrCodeCorruptObj to represent a corrupt object:
b) add a new member 'InnerErr error' to StorageError to hold
   the inner error
- add API status error

tkashem commented:
built on top of https://github.com/kubernetes/kubernetes/pull/124674, it will be rebased when 124674 merges.

/cc @dgrisonnet @stlaz

tkashem committed:
api: add a new field to meta/v1 DeleteOptions

- add a new boolean field
  IgnoreStoreReadErrorWithClusterBreakingPotential to meta/v1 DeleteOptions

- add validation for the new delete option
add validation for the new field in the delete options
ignoreStoreReadErrorWithClusterBreakingPotential

- prevent the pod eviction handler from issuing an unsafe pod delete
prevent the pod eviction handler from enabling the
'ignoreStoreReadErrorWithClusterBreakingPotential' delete option

tkashem committed:
api: run codegen

run 'make update' to code gen for changes in meta/v1 DeleteOptions

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

stlaz reviewed (commented):
First pass.

I really like the idea of passing in the decoder and transformer that wrap the errors, that makes it easier to plug the feature in IMO. I think the embedded store makes sense at this point of feature maturity, I hope it won't make it harder to merge with the current code later.

I am not sure about the side-effect of the error aggregation in listing. I also wonder if it would be possible to handle the Delete() in a more straightforward fashion than with a separate object.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/unsafe_deleter_store.go:
```diff
@@ -0,0 +1,120 @@
+// 'IgnoreStoreReadErrorWithClusterBreakingPotential' to true
+// c) the given store read error represents a corrupt object
+// otherwise, it returns false
+func (s *corruptObjectDeleter) IsCandidateForUnsafeDeletion(err error, options *metav1.DeleteOptions) bool {
```
this method has a strange signature, not helpful IMO. Hiding core logic in a predicate makes it look complex.

Better above in the `Delete` method:

```golang
if storage.IsCorruptObject(err) && options != nil && options.IgnoreStoreReadErrorWithClusterBreakingPotential {
  ...
}
```

tkashem replied:
I was thinking of using the same method signature as `GracefulDeleter` 
```
type CorruptObjectDeleter interface {
	// for a corrupt object we forego graceful deletion and finalizer
	// constraints, and delete the object from the storage
	GracefulDeleter

	IsCandidateForUnsafeDeletion(err error, options *metav1.DeleteOptions) bool
}
```

based on:
https://github.com/kubernetes/kubernetes/blob/78d64904128add672b9f5d48d676c0bdc6a1c2de/staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go#L169-L181

alternatively, we could just define a new interface, same as the  `GracefulDeleter.Delete` method, but add a new parameter  `deleteErr error`, thoughts?
```
type CorruptObjectDeleter interface {
	Delete(ctx context.Context, name string, deleteValidation ValidateObjectFunc, options *metav1.DeleteOptions, deleteErr error) (runtime.Object, bool, error)
}
```

tkashem replied:
oh i see, you are saying to move the precondition inside Delete of store.go, sure I can do that

Jefftree commented:
/remove-sig api-machinery

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,121 @@
+		return admission.NewForbidden(attr, err)
+	}
+	if decision == authorizer.DecisionAllow {
+		klog.Infof("permitting verb %q on %s/%s to user: %q", verb, record.Namespace, record.Name, record.User.GetName())
```
This seems to be verbose. Usually admission plugins don't log or above verbosity 4.

tkashem replied:
I removed the log line

tkashem commented:
/retest

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -90,6 +90,22 @@ type Config struct {
+	// WARNING: Vendor(s) will most likely consider using this option to be
+	// in violation of the support of their product.
+	// The default value is false, and the user must opt in to enable it.
+	AllowUnsafeCorruptObjectDeletion bool
```
- In the future, when the feature gate is removed, this can be used as a knob to enable/disable the feature.
- this also gives us the ability to tweak this feature per resource, in case we decide that we will never enable this feature for certain types.

I can remove this and just use  `utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion)`, thoughts @sttts @stlaz ?

sttts replied:
No strong opinion. But I tend to agree with @stlaz. We can make it configurable when we see the need.

tkashem commented:
/retest

stlaz commented:
https://github.com/kubernetes/enhancements/issues/3926 xref w/ the issue

stlaz commented:
/assign
I am currently a little busy but should be able to review the PR again hopefully in the first half of the next week.

seans3 commented:
/triage accepted

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -271,13 +271,13 @@ func (s *store) Delete(
-	validateDeletion storage.ValidateObjectFunc, cachedExistingObject runtime.Object) error {
-	getCurrentState := s.getCurrentState(ctx, key, v, false)
+	validateDeletion storage.ValidateObjectFunc, cachedExistingObject runtime.Object, transformDecode bool) error {
+	getCurrentState := s.getCurrentState(ctx, key, v, false, transformDecode)
```
`conditionalDelete` runs precondition and admission; the unsafe deletion is also conditional, we still run the admission. I wanted to keep the the current `conditionalDelete` method as is and control deletion with or without `skipTransformDecode`. That seemed more maintainable imo, thoughts?

tkashem commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/dryrun.go:
```diff
@@ -46,7 +46,7 @@ func (s *DryRunnableStorage) Create(ctx context.Context, key string, obj, out ru
 }
 
-func (s *DryRunnableStorage) Delete(ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions, deleteValidation storage.ValidateObjectFunc, dryRun bool, cachedExistingObject runtime.Object) error {
+func (s *DryRunnableStorage) Delete(ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions, deleteValidation storage.ValidateObjectFunc, dryRun bool, cachedExistingObject runtime.Object, deleteOptions storage.DeleteOptions) error {
```
I think so, the preconditions get filled here:
https://github.com/kubernetes/kubernetes/blob/e673417529d9d2d2dd04f28f8437cbe1c403dfbe/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1125-L1128

tkashem commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1113,7 +1127,19 @@ func (e *Store) Delete(ctx context.Context, name string, deleteValidation rest.V
 	}
```
if the unsafe deletion flow is mingled with the graceful deletion flow we have today, it might be harder to follow the code, i think having the logic for unsafe deletion co-located in a separate file makes it more easy to follow. 

A couple of points:
- If we want to know whether graceful Deletion has failed, the unsafe deleter needs to wrap the graceful deleter
- the way i am wiring the unsafe deleter is from the `DeleteResource` endpoint handler. The unsafe deleter is used if and only if the user has enabled the delete option.

tkashem replied:
it's easier to wire the unsafe-deleter from inside if we have it as a separate unit, rather than making the changes inline
```
		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
			if options != nil && ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
				p, ok := r.(rest.CorruptObjectDeleterProvider)
				if !ok {
					// TODO: this is a developer error
					scope.err(fmt.Errorf("ignoreStoreReadErrorWithClusterBreakingPotential is not supported"), w, req)
					return
				}
				r = p.GetCorruptObjDeleter()
			}
		}
```

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/unsafe_deleter_store.go:
```diff
@@ -0,0 +1,194 @@
+	return &errorInterpretingTransformer{Transformer: transformer}
+}
+
+// storeWithUnsafeDelete abstracts unsafe deletion flow for etcd3 storage
```
> This wrapper should be merged with the etcd3 store
> // once the `AllowUnsafeMalformedObjectDeletion` feature goes stable

same as above, there is precedence for introducing a feature through extension, as opposed to inline modification. For example, the dryRun implementation.

Having said that, if folks prefer inline changes, i will oblige :)

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -461,7 +461,23 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
+
+	if c.AllowUnsafeCorruptObjectDeletion {
+		transformer = etcd3.NewErrorInterpretingTransformer(transformer)
+		decoder = etcd3.NewErrorInterpretingDecoder(decoder)
```
Yes, I believe so.
Unfortunately inside the transformer or the decoder, we don't have access to the key of the object, so the error created from the wrapped transformer/decoder does not have any information that identifies the object.

tkashem replied:
also, `InternalError` is not limited to transformation error only:

https://github.com/kubernetes/kubernetes/blob/190eb057f4eaac1db47b54f0d8232eae6ebd1ff2/staging/src/k8s.io/apiserver/pkg/storage/errors.go#L171-L173

`InternalError` is thrown from precondition check:
https://github.com/kubernetes/kubernetes/blob/190eb057f4eaac1db47b54f0d8232eae6ebd1ff2/staging/src/k8s.io/apiserver/pkg/storage/interfaces.go#L136-L147

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -782,7 +790,15 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 				recordDecodeError(s.groupResourceString, string(kv.Key))
-				return err
+				switch {
+				case opts.AggregateCorruptObjFn != nil:
```
yeah, let me look into it.

tkashem commented:
> FAIL: Test_UncertainDeviceGlobalMounts

/retest

dgrisonnet commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -66,7 +84,10 @@ resources:
     - secrets
     providers:
-    - identity: {}
+    - aesgcm:
+        keys:
+        - name: key1
+          secret: dXBlcnByZGVscHJkZWxwCg==
```
This is the `identityConfigYAML`, why are we changing the provider to `aesgcm` instead of using `aesGCMConfigYAML`?

enj replied:
(this seems unexpected to me)

tkashem replied:
yes, it's fixed now. the test configures encryption with `aesGCMConfigYAML` and then it replaces the config entirely with `aesCBCConfigYAML` resulting in a `no matching prefix found` error

dgrisonnet commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+					return got.Status().Reason == metav1.StatusReasonInternalError &&
+						strings.Contains(got.Status().Message, "StorageError: corrupt object")
+				},
+				corruptObjGet:          wantAPIStatusError{reason: metav1.StatusReasonInternalError},
```
I think we should add a test that verifies that failed LIST return a list of corrupted objects. Maybe in a separate test? Or that already covered by unit tests?

tkashem replied:
yes, I have a separate integration test for that: https://github.com/kubernetes/kubernetes/pull/127513/commits/c5d8d5ae850512ae340a25c32d611057da4d531d, please take a look :)

tkashem commented:
/cc @wojtek-t

tkashem commented:
/label api-review

it implements KEP-3927: https://github.com/kubernetes/enhancements/pull/3927
The difference is, instead of using a `IgnoreStoreReadErrorWithClusterBreakingPotential bool`, we are using `IgnoreStoreReadErrorWithClusterBreakingPotential *bool`

tkashem commented:
/remove-label api-review

tkashem commented:
/label api-review

it implements KEP-3926: https://github.com/kubernetes/enhancements/pull/3927
The difference is, instead of using a IgnoreStoreReadErrorWithClusterBreakingPotential bool, we are using IgnoreStoreReadErrorWithClusterBreakingPotential *bool

liggitt commented on staging/src/k8s.io/api/testdata/HEAD/policy.v1.Eviction.json:
```diff
@@ -53,6 +53,7 @@
       "dryRunValue"
-    ]
+    ],
+    "ignoreStoreReadErrorWithClusterBreakingPotential": true
```
interesting... should the eviction handler prevent setting this option to true in a pod evict request?

tkashem replied:
I think the eviction handler should prevent it, maybe there is a better place to do it, i added the check inside `EvictionREST` https://github.com/kubernetes/kubernetes/pull/127513/commits/5e1885af6d47f460c9c48e8ebe1048b5f849038b

deads2k replied:
Seems weird that the testdata here added it.  Have you worked out why yet?

tkashem replied:
I thought the testdata adds a non default value for every field for api roundtrip testing, and eviction request has a field `deleteOptions *metav1.DeleteOptions`
https://github.com/kubernetes/kubernetes/blob/2bb886ce2a6ae5a83031efd225e8a17dd7ecba03/staging/src/k8s.io/api/policy/v1/types.go#L213-L223

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -357,3 +358,28 @@ func isValidConditionReason(value string) []string {
+// ignoreStoreReadErrorWithClusterBreakingPotential is set
+func ValidateIgnoreStoreReadError(fldPath *field.Path, options *metav1.DeleteOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if ignore := options.IgnoreStoreReadErrorWithClusterBreakingPotential; ignore != nil && *ignore {
```
```suggestion
	if ignore := ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false); ignore {
```

tkashem replied:
done

liggitt commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -61,6 +61,16 @@ const (
 	AnonymousAuthConfigurableEndpoints featuregate.Feature = "AnonymousAuthConfigurableEndpoints"
 
+	// owner: @stlaz @tkashem @dgrisonnet
+	// alpha: v1.32
```
drop the alpha release comment, it's included in the versioned registration below

tkashem replied:
yes, dropped

liggitt commented:
/assign @deads2k 
for API bits

/assign @enj 
for kms integration test bits

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -234,6 +234,23 @@ type Store struct {
+	// a) the data (associated with the resource being deleted) retrieved
+	// from the storage failed to transform properly (eg. decryption failure)
+	// b) the data (associated with the resource being deleted) failed to
+	// decode properly (eg. corrupt data)
+	// it will disregard these errors, bypass the finalzer constraints,
```
```suggestion
	// decode properly (eg. corrupt data).
	// It will disregard these errors, bypass the finalzer constraints,
```

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -234,6 +234,23 @@ type Store struct {
+	// it will disregard these errors, bypass the finalzer constraints,
+	// deletion hook(s) and go ahead with the deletion of the object.
+	//
+	// WARNING: This may break the cluster if the resource has
+	// dependencies. Use when the cluster is broken, and there is no
+	// other viable option to repair the cluster.
```
```suggestion
	// WARNING: This may break the cluster e.g. if the resource has
	// dependencies. Use when the cluster is broken, and there is no
	// other viable option to repair the cluster.
```

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -234,6 +234,23 @@ type Store struct {
+	// other viable option to repair the cluster.
+	//
+	// TODO: it's located here for convenience so we can wire it
+	// from inside the handler.DeleteResource function
```
what is the TODO? Is it wrong here? Where should it go?

tkashem replied:
ideally i would like to do wrap the `registry.Store` with the following:
```
type corruptObjectDeleter struct {
    *registry.Store
}
```

and pass an instance of `corruptObjectDeleter` to api installer, but REST factory functions for all types directly instantiates `registry.Store{...}`

https://github.com/kubernetes/kubernetes/blob/f1e447b9d32ac325074380d239370cde02a6dbf7/pkg/registry/core/endpoint/storage/storage.go#L36-L55

So I am holding the `CorruptObjDeleter` instance inside `registry.Store` so I can access it from 
https://github.com/kubernetes/kubernetes/blob/f1e447b9d32ac325074380d239370cde02a6dbf7/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L50

enj commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+	}
+}
+
+type listCorruptObjectTestRunner struct {
+	featureEnabled bool
+	// secrets that are created before encryption breaks
+	secrets []string
+	// whether encryption broke after the config change
+	encryptionBrokenFn func(t *testing.T, got apierrors.APIStatus) bool
+	// what we expect for LIST on the corrupt objects after encryption has broken
+	listAfter verifier
+}
+
+func (testrunner listCorruptObjectTestRunner) Run(t *testing.T) {
```
Same comment regarding the struct and the Run method.

tkashem replied:
yes, done

enj commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+		user, namespace, verbs[0], "", schema.GroupResource{Resource: "secrets"}, true)
+}
+
+func TestListCorruptObjects(t *testing.T) {
```
Add a description of what the test is trying to assert.

tkashem replied:
comment added

enj commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+	corrupObjFinalGet verifier
+}
+
+func TestAllowUnsafeMalformedObjectDeletionFeature(t *testing.T) {
```
Add a description of what the test is trying to assert.

tkashem replied:
comment added

enj commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -560,6 +560,23 @@ type DeleteOptions struct {
+	// removes the object from the storage.
+	// WARNING: This will break the cluster if the resource has any
+	// dependencies. Use only if you REALLY know what you are doing.
+	// WARNING: Vendor(s) will most likely consider enablement of this
+	// option to be in violation of the support of their product.
+	// The default value is nil, and the user must opt in to enable it
```
I haven't looked at the KEP / implementation, but can the cluster operator disable this functionality via a CLI flag?  Given this warning, should they be able to?

tkashem replied:
for alpha, the feature gate is atomic, it's disabled by default; and the cluster operator can disable it via the --feature-flag option.
The KEP does not address whether the cluster operator can disable the feature via a server run option, I think we should add this as a beta criteria, I would feel safer if the operator has a way to disable this feature via CLI once the feature gate is removed. Thoughts @stlaz?

tkashem replied:
update: based on the discussion on slack https://kubernetes.slack.com/archives/C0EG7JC6T/p1729678373175029 we are not providing a CLI option to disable the feature, we will have docs with example VAP that shows how to disable the feature.

enj reviewed (changes requested):
I only looked at the integration tests.  Most of the comments are around making the test style match the normal pattern we follow.

tkashem commented on test/integration/controlplane/transformation/transformation_test.go:
```diff
@@ -276,14 +278,19 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
+	}
+	if e.corruptObjDeleteEnabled {
+		options = append(options, []string{
+			"--authorization-mode=RBAC"}...)
```
the admission plugin used for unsafe delete is enabled by default (but it is active only when the feature gate is enabled). if folks have no concern with it, then all i need is to enable RBAC for the new tests.

For existing tests, RBAC is not enabled today, I pulled the refactor commit into a separate PR to make review easier: https://github.com/kubernetes/kubernetes/pull/128191

tkashem commented:
/retest

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,123 @@
+		Name:            attr.GetName(),
+		APIGroup:        attr.GetResource().Group,
+		APIVersion:      attr.GetResource().Version,
+		Resource:        attr.GetResource().Resource,
```
We drop subresource?

tkashem replied:
I changed the plugin to act on resource only request.

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,123 @@
+	if !utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion) {
+		return nil
+	}
+	if attr.GetOperation() != admission.Delete || attr.GetOperationOptions() == nil {
```
do we have to check for resource request? Is there a chance to get options on a non-resource endpoint?

tkashem replied:
i am not sure, but i changed the plugin to act on a resource only request, no sub-resource or non-resource endpoint

sttts commented on plugin/pkg/admission/allowunsafedelete/admission_test.go:
```diff
@@ -0,0 +1,231 @@
+		nil,                           // this plugin should never inspect the object
+		nil,                           // old object
+		schema.GroupVersionKind{},     // this plugin should never inspect kind
+		"",                            // this plugin should never inspect namespace
```
not true comment, right?

tkashem replied:
the plugin simply passes it to the authorizer, i updated the comment

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+var _ rest.GracefulDeleter = &corruptObjectDeleter{}
+
+// NewCorruptObjectDeleter returns a deleter that can perform unsafe deletion
+// of corrupt objects, it can be invoked to delete corrupt objects only after
```
"it can be ... only", not sure I understand the "only". If this is about the user, then be explicit. If it is about the wired code, then it is not "it can" but "it will".

tkashem replied:
comment updated

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+// b) failure to decode the object
+//
+// NOTE: it skips precondition checks, finalizer constraints, and any
+// post deletion hooks.
```
what is a post deletion hook?

tkashem replied:
It's not user facing, but is a configurable property of the registry.

https://github.com/kubernetes/kubernetes/blob/d8919343526597e0788a1efe133c70d9a0c07f69/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L198-L200

sttts replied:
better use the same terms: after delete func.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+// NOTE: it skips precondition checks, finalizer constraints, and any
+// post deletion hooks.
+//
+// WARNING: This may break the cluster if the resource being deleted has dependencies
```
```suggestion
// WARNING: This may break the cluster if the resource being deleted has dependencies.
```

tkashem replied:
fixed

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+	Storage storage.Interface
+}
+
+// Delete performs an unsafe deletion of the given resource from the storage
```
```suggestion
// Delete performs an unsafe deletion of the given resource from the storage.
```

sttts replied:
same everywhere else where a comment is a complete sentence, in particular multi-sentence comments.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,146 @@
+	}
+	obj := s.NewFunc()
+
+	// TODO: what if Get returns the object from cache, in this case the
+	// apiserver must restart for unsafe deletion to work
+	err = s.Storage.Get(ctx, key, storage.GetOptions{}, obj)
```
why this check in the first place? Why don't we trust the request? It is authorized after all. If we want to be careful, I would rather expect that we try a normal delete and fall back to the special form.

sttts replied:
I see in the code below what the get is here, namely to create the RV precondition. This defensive programming is horrible to reason about 🤔

tkashem replied:
The primary goal of get operation is to make sure that the resource being deleted is corrupt, if the error from get represents a corrupt object, only then we proceed with the unsafe deletion. (we want to limit unsafe delete to corrupt objects only)

The precondition pins to the RV from the get operation, so the unsafe delete fails if the object gets recreated or updated in between. The RV precondition covers an unlikely edge case, I can remove it if it makes the delete flow hard to reason about.

Today the default behavior of normal deletion is no precondition and hence the object is deleted at HEAD. We can keep the same semantics for unsafe delete.

tkashem replied:
empty resource version in the get options forces the watch cache to serve from the underlying storage

https://github.com/kubernetes/kubernetes/blob/b95484e330d7303e67e13b145e0f174c6c1c1da7/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L703-L708

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+	// keep the admission
+	if err := s.Storage.Delete(ctx, key, out, preconditions, storage.ValidateObjectFunc(deleteValidation), nil, storageOpts); err != nil {
+		if storage.IsNotFound(err) {
+			// the DELETE succeeded, but we don't have the object sine it's
```
```suggestion
			// the DELETE succeeded, but we don't have the object since it's
```

tkashem replied:
fixed

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,145 @@
+		return nil, false, apierrors.NewInternalError(fmt.Errorf("expected %s to be enabled", optionName))
+	}
+
+	qualifiedResource := s.qualifiedResourceFromContext(ctx)
```
move that down where it is used.

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,228 @@
+	return &corruptObjectDeleter{
+		Interface: delegate,
+		// TODO: make it configurable?
+		aggregateErrMaxCount: 100,
```
what's the idea to have a maximum? Why not go the easy path and return all errors?

tkashem replied:
To prevent from having a large message, from the KEP discussion: https://github.com/kubernetes/enhancements/pull/3927/files#r1345915347

> why not go the easy path and return all errors

The KEP mentions returning malformed object(s), we limit unsafe deletion to malformed/corrupt objects only, so the admin seeing other errors may be confusing?

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -313,8 +313,24 @@ type ListOptions struct {
+	// - key: identifies the object in the underlying storage
+	// - err: the error that occurred while retrieving the given object from the storage
+	// - done: if true, the list operation will abort, otherwise, it will continue
+	AggregateErrFn func(key string, err error) (done bool)
```
Can't we be super explicit what this does instead of allowing an arbitrary function? This method here is a bad abstraction anyway, let's not pretend it isn't.

tkashem replied:
this abstraction has been removed

tkashem commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+
+func TestAllowUnsafeMalformedObjectDeletionFeature(t *testing.T) {
+	tests := []struct {
+		runner allowUnsafeMalformedObjectDeletionTestRunner
```
yes, moved inline

tkashem commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("%s/%t", string(genericfeatures.AllowUnsafeMalformedObjectDeletion), tc.runner.corruptObjDeleteEnabled), func(t *testing.T) {
+			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.AllowUnsafeMalformedObjectDeletion, tc.runner.corruptObjDeleteEnabled)
+			tc.runner.Run(t)
```
yes, moved inline

tkashem commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,123 @@
+	return nil
+}
+
+// Validate makes sure that all containers are set to always pull images
```
fixed

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented on staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go:
```diff
@@ -200,3 +200,23 @@ func AdmissionToValidateObjectDeleteFunc(admit admission.Interface, staticAttrib
+
+// WithCorruptObjDeleter wraps the given rest.GracefulDeleter that is used for normal
+// deletion flow with an unsafe deleter to facilitate deletion of corrupt object
+func WithCorruptObjDeleter(deleter GracefulDeleter, opts *metav1.DeleteOptions) GracefulDeleter {
```
The constraints I found are:
- the unsafe deleter need to be initialized inside 
```
-1631,9 +1648,17 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {

	if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
		e.CorruptObjDeleter = NewCorruptObjectDeleter(e)
	}

	return nil
}
```

and also the fact that `regisrty.Store` is instantiated  `NewRegistry` function of every type  https://github.com/kubernetes/kubernetes/pull/127513#discussion_r1809061069

sttts replied:
With https://github.com/tkashem/kubernetes/pull/1, I think I am good with the general direction.

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,138 @@
+	// if the feature is disabled, this plugin, although
+	// enabled, should not be active
+	if !utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion) {
+		return nil
```
this means we don't refuse requests with the flag set, we ignore the flag. Just to be sure that's what we want. A oneliner go doc here wouldn't harm to call that out.

tkashem replied:
comment added

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,138 @@
+	if !ok {
+		return fmt.Errorf("expected an option of type: %T, but got: %T", &metav1.DeleteOptions{}, attr.GetOperationOptions())
+	}
+	if ignore := options.IgnoreStoreReadErrorWithClusterBreakingPotential; ignore == nil || !*ignore {
```
there is `ptr.Deref` for this kind of checks.

tkashem replied:
yes, i have been using it

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,138 @@
+	decision, reason, err := p.authz.Authorize(ctx, record)
+	if err != nil {
+		err = fmt.Errorf("error while checking permission for %q, %w", verb, err)
+		klog.V(1).ErrorS(err, "failed to authorize")
```
wondering, don't we have the structured logger from the context? This kind of global klog is deprecated I believe.

tkashem replied:
sure, added context based logging

sttts commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,138 @@
+		return nil
+	}
+
+	return admission.NewForbidden(attr, fmt.Errorf("user not permitted to do %q, reason: %s", verb, reason))
```
```suggestion
	return admission.NewForbidden(attr, fmt.Errorf("not permitted to do %q, reason: %s", verb, reason))
```

tkashem replied:
fixed

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -357,3 +360,28 @@ func isValidConditionReason(value string) []string {
+// ignoreStoreReadErrorWithClusterBreakingPotential is enabled
+func ValidateIgnoreStoreReadError(fldPath *field.Path, options *metav1.DeleteOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if ignore := ptr.Deref[bool](options.IgnoreStoreReadErrorWithClusterBreakingPotential, false); ignore {
```
```suggestion
	if ignore := ptr.Deref[bool](options.IgnoreStoreReadErrorWithClusterBreakingPotential, false); !ignore {
		return allErrs
	}
```

Return early

tkashem replied:
done

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -357,3 +360,28 @@ func isValidConditionReason(value string) []string {
+	allErrs := field.ErrorList{}
+	if ignore := ptr.Deref[bool](options.IgnoreStoreReadErrorWithClusterBreakingPotential, false); ignore {
+		if len(options.DryRun) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, ignore, "ignoreStoreReadErrorWithClusterBreakingPotential and dryRun cannot be both set"))
```
`ignoreStoreReadErrorWithClusterBreakingPotential` is part of the path and with that visible in the error. `"cannot be set together with .dryRun"` should be enough.

tkashem replied:
fixed.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go:
```diff
@@ -200,3 +200,23 @@ func AdmissionToValidateObjectDeleteFunc(admit admission.Interface, staticAttrib
+// WithCorruptObjDeleter wraps the given rest.GracefulDeleter that is used for normal
+// deletion flow with an unsafe deleter to facilitate deletion of corrupt object
+func WithCorruptObjDeleter(deleter GracefulDeleter, opts *metav1.DeleteOptions) GracefulDeleter {
+	if opts == nil {
+		return deleter
+	}
+	if ignore := opts.IgnoreStoreReadErrorWithClusterBreakingPotential; ignore == nil || !*ignore {
+		return deleter
+	}
```
all this is not needed if you inline the code below directly into `DeleteResource`:
```golang
		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
			if ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
				if p, ok := r.(rest.CorruptObjectDeleterProvider); !ok {
					scope.err(fmt.Errorf("ignoreStoreReadErrorWithClusterBreakingPotential is not supported"), w, req)
					return
				} else {
					r = p.GetCorruptObjDeleter()
				}
			}
		}
```
Or in other words: pulling logic apart and making it apparently reusable makes it much easier to reason about. This function here is a noop in certain case, but that's totally invisible from the function name and signature.

Has it been called out (maybe in the KEP) that we ignore the options flags for storages that don't support it?

tkashem replied:
the KEP does not mention this, it's a developer error if the feature is enabled, and `r.(rest.CorruptObjectDeleterProvider)` fails.

sftim commented:
This needs a real release note; it's not mergeable as is @tkashem.

tkashem commented:
@sttts I refactored list error aggregation at these commits, ptal https://github.com/kubernetes/kubernetes/pull/127513/commits/403fd552c15b2927c2bdc54b29f3a0f5a10b4252
https://github.com/kubernetes/kubernetes/pull/127513/commits/dd049c2e7571ad9a0c38eb929d146e5ac589a1ba

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -768,7 +768,15 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 			if err != nil {
-				return storage.NewInternalError(fmt.Errorf("unable to transform key %q: %w", kv.Key, err))
+				switch {
+				case opts.AggregateCorruptObjFn != nil:
```
yes, i have removed the list option, and added it as a member of the store

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -122,6 +123,18 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
+					scope.err(fmt.Errorf("ignoreStoreReadErrorWithClusterBreakingPotential is not supported"), w, req)
+					return
+				}
+				r = p.GetCorruptObjDeleter()
```
@sttts should we say `internal error, ignoreStoreReadErrorWithClusterBreakingPotential is not initialized"` instead?

Also, do we want to add?
```
deleter := p.GetCorruptObjDeleter()
if deleter == nil {
     scope.err(fmt.Errorf("internal error, ignoreStoreReadErrorWithClusterBreakingPotential is not initialized"), w, req)
     return
}
```

sttts replied:
hot take: should we add it to the graceful deleter interface?

sttts replied:
Only downsides: ppl might not notice that they don't support the new GracefulDeleter interface anymore.

sttts replied:
In theory you could check before the closure that `r` has the interface. That non-nil is return by the methods you can specify in the interface.

tkashem replied:
fixed it

tkashem commented:
/retest

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -560,6 +560,20 @@ type DeleteOptions struct {
 	// +listType=atomic
 	DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,5,rep,name=dryRun"`
+
+	// IgnoreStoreReadErrorWithClusterBreakingPotential, if set to true,
```
nit: the style here in the struct is to not mention the field name as first word.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -560,6 +560,20 @@ type DeleteOptions struct {
+	// will trigger an unsafe deletion of the resource in case the
+	// normal deletion flow fails with a corrupt object error.
+	// A resource is considered corrupt if it can not be retrieved from
+	// the underlying storage because of a) its data can not be transformed
```
"transform" is not a term in the API, is it? At least I would mention "e.g. decrypted ....".

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -560,6 +560,20 @@ type DeleteOptions struct {
+	// successfully, or b) it fails to decode into an object.
+	// NOTE: unsafe deletion ignores finalizer constraints, skips
+	// precondition checks, and removes the object from the storage.
+	// WARNING: This will break the cluster if the resource has any
+	// dependencies. Use only if you REALLY know what you are doing.
+	// The default value is nil, and the user must opt in to enable it
```
"dependency" is not an API term. "nil" is Go level, "false" is API level.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,118 @@
+	// use the storage implementation directly, bypass dryrun implementation
+	storageBackend := d.store.Storage.Storage
+	// TODO: what if Get returns the object from cache, in this case the
+	// apiserver must restart for unsafe deletion to work
```
this comment feels important. Where is the wrapping with the watch cache happening? Are we sure in k/k at least that this cannot come from cache? @p0lyn0mial might know.

tkashem replied:
okay, i looked it up, if you don't specify a resource version then it will be served from the underlying storage. I will update the comment.

https://github.com/kubernetes/kubernetes/blob/b95484e330d7303e67e13b145e0f174c6c1c1da7/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L703-L708

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,118 @@
+		}
+	}
+
+	klog.FromContext(ctx).V(1).Info("Going to perform unsafe object deletion", "object", klog.KRef(genericapirequest.NamespaceValue(ctx), name))
```
the logger in the context does have the gvr, right?

tkashem replied:
it just returns an `ObjectRef` which has a `String` method
```
// KRef returns ObjectRef from name and namespace
func KRef(namespace, name string) ObjectRef {
	return ObjectRef{
		Name:      name,
		Namespace: namespace,
	}
}
```

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/admission.go:
```diff
@@ -219,6 +221,24 @@ func (a *AdmissionOptions) Validate() []error {
+		}
+		if disablePlugins.Has(name) {
+			errs = append(errs, fmt.Errorf("the plugin %q must not be disabled when feature %q is enabled", name, name))
+		}
```
do we have similar checks for other feature gates? I thought that configuring admission is highly discouraged, but technically fine to do.

tkashem replied:
we are moving admission to the REST layer

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -99,12 +100,43 @@ type objState struct {
+// DefaultListErrorAggregatorFactory returns the default list error
+// aggregator that maintains backward compatibility, which is abort
+// the list operation as soon as it encounters the first error
+func DefaultListErrorAggregatorFactory() ListErrorAggregator { return &abortOnFirstError{} }
```
I still think this is all 2 levels too meta. A factory for a leaking abstraction.

tkashem replied:
we need to switch to corrupt object error aggregation when the feature is enabled https://github.com/kubernetes/kubernetes/pull/127513/commits/d09040330a93994592a100899bf8aca208448f54

The other option is to make these changes inline inside store.go with `if feature enabled {} else {}` which i wanted to avoid. Also when the feature gate is removed we won't need this factory, we will have just one type of aggregator. Thoughts?

deads2k replied:
@sttts were you satisfied on this thread?

sttts commented:
Overall shape is fine now, with https://github.com/kubernetes/kubernetes/pull/127513#discussion_r1816692067 resolved (that one sounded worrying).

https://github.com/kubernetes/kubernetes/pull/127513#discussion_r1816712339 is left, but it's rather cosmetics and taste.

I think it is ready for @deads2k to take a look at the API and its semantics.

tkashem commented:
/retest

tkashem commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -560,6 +560,23 @@ type DeleteOptions struct {
+	// b) the data (associated with the resource being deleted) failed to
+	// decode properly (eg. corrupt data)
+	// it will disregard these errors, bypass the finalzer constraints,
+	// deletion hook(s) and go ahead with the deletion flow.
```
going to mark this as resolved since the description has already gone through a number of revisions, and it is still waiting API review.

deads2k commented on pkg/kubeapiserver/options/plugins.go:
```diff
@@ -91,6 +92,7 @@ var AllOrderedPlugins = []string{
 	certsubjectrestriction.PluginName,       // CertificateSubjectRestriction
 	defaultingressclass.PluginName,          // DefaultIngressClass
 	denyserviceexternalips.PluginName,       // DenyServiceExternalIPs
+	allowunsafedelete.PluginName,            // AllowUnsafeMalformedObjectDeletion
```
Seems like this should be very early in the chain since it fails all the others are not run.

tkashem replied:
I moved it next to `NamespaceAutoProvision`

tkashem replied:
no longer applicable, resolving it

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -234,6 +234,21 @@ type Store struct {
+	// dependencies. Use when the cluster is broken, and there is no
+	// other viable option to repair the cluster.
+	//
+	// TODO: it's located here for convenience so we can wire it
+	// from inside the handler.DeleteResource function.
```
If it's a TODO, what will this eventually change to?

tkashem replied:
I don't think it's necessary to change , `registry.Store` is used as a container for configuring the api handler from the storage, so the `Store` is a convenient holder for this, I removed the comment.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,139 @@
+// 'IgnoreStoreReadErrorWithClusterBreakingPotential' in the delete options.
+func (d *corruptObjectDeleter) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, opts *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if opts == nil || !ptr.Deref[bool](opts.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
+		// this is a developer error, the unsafe-deleter is wired in the rest
+		// layer only when the option is enabled, should we return an error instead?
+		return d.store.Delete(ctx, name, deleteValidation, opts)
```
Anything preventing us from returning an error here?  A 500 error seems appropriate here to ensure it doesn't happen.

tkashem replied:
yes, i am in favor of returning an error here, thanks

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,139 @@
+	// keep the admission
+	if err := storageBackend.Delete(ctx, key, out, preconditions, storage.ValidateObjectFunc(deleteValidation), nil, storageOpts); err != nil {
+		if storage.IsNotFound(err) {
+			// the DELETE succeeded, but we don't have the object since it's
+			// not retrievable from the storage, so we send a nil object
+			return nil, false, nil
```
I haven't finished reviewing so it may already exist, but we need to see what currently well behaved informers do with this.  In particular, what's in my head is the resourceVersion on a watch.  I remember there was (and still is I think) an issue where events can be deleted in the same resourceVersion.  I'd like to find a way prove that the deletion has a unique resourceVersion.

tkashem replied:
as far as response to the delete request, for a normal deletion flow, we return the old object with a more recent revision:
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L368

so for a watcher watching this object, it will see these events:
- a) `ADDED`, RV: x 
- b) `DELETED` RV: y
for both a and b, it's the same object in `origState.data`, but y > x because of `txnResp.Revision`

However for unsafe deletion, since we can not transform or decode, we return a `nil` object to the response of the delete request.

so for a watcher watching this object, it will see these events:
- a) `ADDED`, RV: x 
- b) instead of a `DELETED` event, it will see an `ERROR` event because of transform or decode failure
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L681-L697

https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L564-L569

https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L636-L646

`prepareObjs` returns an error like this:
```
failed to prepare current and previous objects: data from the storage is not transformable: no matching prefix found
```
and it causes the watcher to terminate with an `watch.Error` event, and the reflector sees this as an internal error and retries the watch
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/client-go/tools/cache/reflector.go#L488-L490

> I remember there was (and still is I think) an issue where events can be deleted in the same resourceVersion

not sure what events are you referring to here, events as in event resource?

tkashem replied:
> b) instead of a DELETED event, it will see an ERROR event because of transform or decode failure

we can try to construct an empty object, fill it with the namespace, name, and the revision after delete, this will result in construction of a watch.Deleted event with a partial object. This is hacky, and other implications, what if a workload is generating keys for the store based on other attributes of the object like from the labels or annotations

tkashem replied:
@deads2k I have a solution at https://github.com/kubernetes/kubernetes/pull/127513/commits/46c3fdc4eeda8d3ffb1839fff196c5fece75066d let me know if this is palatable, I can bring it up for discussion.

sttts replied:
What happens if `handleUnsafeDeleteEventError` is not called yet because the client library is old in some component? Will an informer eventually do the right thing? Is there a way to prove that?

tkashem replied:
@sttts we have opted to implement option A from https://github.com/kubernetes/enhancements/pull/4949

https://github.com/kubernetes/kubernetes/pull/127513/commits/0f4dd0bd5c93865b0d46cbc4710e31a313f5cc89 does not introduce any change in the client, so existing/old clients would work just fine.

does this answer the question you have?

sttts replied:
yes, with option A it does. The referenced commit above is not part of the PR.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/options/admission.go:
```diff
@@ -219,6 +221,24 @@ func (a *AdmissionOptions) Validate() []error {
+		if !registeredPlugins.Has(name) {
+			errs = append(errs, fmt.Errorf("the plugin %q must be registered, and enabled by default", name))
+		}
+		if disablePlugins.Has(name) {
```
What is ensuring it is enabled?    Disabled is just a force disabled list and since this is being set a DefaultOff, what turns it on?

tkashem replied:
the plugin is enabled by default, but it is active only when the feature is enabled. I enabled the plugin by removing it from `DefaultOffAdmissionPlugins` :
```
modified   pkg/controlplane/apiserver/samples/generic/server/admission.go
 
 // DefaultOffAdmissionPlugins get admission plugins off by default for kube-apiserver.
@@ -46,6 +47,7 @@ func DefaultOffAdmissionPlugins() sets.Set[string] {
 		ctbattest.PluginName,                 // ClusterTrustBundleAttest
 		certsubjectrestriction.PluginName,    // CertificateSubjectRestriction
 		validatingadmissionpolicy.PluginName, // ValidatingAdmissionPolicy, only active when feature gate ValidatingAdmissionPolicy is enabled
+		unsafedeletepolicy.PluginName,        // AllowUnsafeMalformedObjectDeletion, only active when feature gate AllowUnsafeMalformedObjectDeletion is enabled
 	)
```

tkashem replied:
no longer relevant, resolving it.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/errors.go:
```diff
@@ -82,30 +84,45 @@ func NewUnreachableError(key string, rv int64) *StorageError {
-		AdditionalErrorMsg: msg,
+		Code: ErrCodeTimeout,
+		Key:  key,
+		err:  errors.New(msg),
```
I'm not entirely clear on why this needed to change.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/errors.go:
```diff
@@ -82,30 +84,45 @@ func NewUnreachableError(key string, rv int64) *StorageError {
 func (e *StorageError) Error() string {
-	return fmt.Sprintf("StorageError: %s, Code: %d, Key: %s, ResourceVersion: %d, AdditionalErrorMsg: %s",
-		errCodeToMessage[e.Code], e.Code, e.Key, e.ResourceVersion, e.AdditionalErrorMsg)
+	return fmt.Sprintf("StorageError: %s, Code: %d, Key: %s, ResourceVersion: %d, err: %v",
```
This change seems capricious.  Why change `AdditionalErrorMsg` to `err`?

tkashem replied:
sure, it stays in the message as is, I have a separate PR for this commit https://github.com/kubernetes/kubernetes/pull/128056

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,269 @@
+	// TODO: right now any error is deemed as undecodable, in
+	// the future, we can apply some filter, if need be.
+	if err := d.Decoder.Decode(value, objPtr, rev); err != nil {
+		return &corruptObjectError{err: err, errType: undecodable}
```
if we know the RV, shouldn't we include it?

tkashem replied:
yes, we have the RV for the decoder, but not for the transformer `TransformFromStorage(ctx context.Context, data []byte, dataCtx value.Context`. We don't have the key in either of the two. So the error management is a little bit convoluted now:

I have a TODO to streamline error management in the storage layer:
- a) remove `InternalError`, and use `StorageError` instead (use a new code `ErrCodeInternal`)
- b) when creating a `StorageError`, do best effort to populate revision and key
- c) use errors.As() for error handling

may be we can add it as a beta criteria?

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -270,7 +303,7 @@ func (s *store) Delete(
 		return fmt.Errorf("unable to convert output object to pointer: %v", err)
 	}
 
-	skipTransformDecode := false
```
this is blowing my mind right now.  We already had this and hardcoded it false?

tkashem replied:
no, we added this as a separate refactor PR: https://github.com/kubernetes/kubernetes/pull/127778

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -912,8 +956,10 @@ func (s *store) getState(ctx context.Context, kv *mvccpb.KeyValue, key string, v
 		if skipTransformDecode {
-			// be explicit that we don't have the object
-			state.obj = nil
+			state.obj = reflect.New(v.Type()).Interface().(runtime.Object)
+			if err := s.versioner.UpdateObject(state.obj, uint64(state.rev)); err != nil {
+				return nil, err
+			}
```
I don't understand this change.  Why is our `getState` function  mutating an object?  The `state.rev` returned from here already has the RV for usage elsewhere, right?

tkashem replied:
If I want to run a precondition check with an RV on an unsafe delete, i need an empty object with an RV set. This does not impact the normal deletion flow though.

tkashem replied:
it's used here https://github.com/kubernetes/kubernetes/pull/127513/commits/b1e40fd2e7bd4649686314ad2e4a6259319fe3d2#diff-592aa733e64bdba1062e9957a7d5b0f4124e7f8102ecb4e4278c7d8215e21af6R116-R119

I don't think it's strictly necessary to close this edge case, i can drop this commit if you want.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -464,7 +464,20 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
+		listErrAggregatorFactory = etcd3.CorruptObjErrAggregatorFactory(100)
+	}
+
+	store := etcd3.New(client, c.Codec, newFunc, newListFunc, c.Prefix, resourcePrefix, c.GroupResource, transformer, c.LeaseManagerConfig, decoder, versioner, listErrAggregatorFactory)
```
why take `listErrAggregatorFactory` instead of setting it in `etcd3.newStore` based on the featuregate?

tkashem replied:
all the initialization bits - wrapping the decoder and the transformer are also here, inside the factory.

tkashem replied:
yes, done

deads2k commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -54,6 +54,15 @@ const (
 	// Allows us to enable anonymous auth for only certain apiserver endpoints.
 	AnonymousAuthConfigurableEndpoints featuregate.Feature = "AnonymousAuthConfigurableEndpoints"
 
+	// owner: @stlaz @tkashem @dgrisonnet
```
it's uncommon for the owners to not be the PR authors, these guy acknowledged?

stlaz replied:
It's ok to remove me from here

tkashem replied:
@stlaz  since you acked, i would like to keep you as an owner, is that okay?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -357,3 +360,31 @@ func isValidConditionReason(value string) []string {
+		return allErrs
+	}
+
+	if len(options.DryRun) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, true, "cannot be set together with .dryRun"))
+	}
```
This one I'm unsure of. If I were a user trying to figure out if I could make this type of delete call, I can't?

tkashem replied:
for a corrupt object, a `GET` will return an error saying the object is corrupt. The KEP mentions the workflow:
- a) LIST returns the corrupt objects with etcd keys
- b) the user issues unsafe delete for these corrupt objects

Today, the user can send a delete with dryRun option for the corrupt object, and it will return the same error as in `a`. 
The first thing the deletion flow does is get the object from the storage, and it will result in the same error as in `a`, so we would not be able to produce more information with dryRun + ignoreStoreReadErrorWithClusterBreakingPotential, thoughts?

deads2k replied:
Still not convinced about the dryrun being unreasonable, but I'm willing to defer th echoice.

deads2k commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,141 @@
+	// if we are here, IgnoreStoreReadErrorWithClusterBreakingPotential
+	// is set to true in the delete options, the user must have permission
+	// to do 'delete-ignore-read-errors' on the given resource.
+	const verb = "delete-ignore-read-errors"
```
we probably want the unsafe bit emphasized here because this will be the point where someone might grant access.

tkashem replied:
yes, changed it to `unsafe-delete-ignore-read-errors`

deads2k commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,141 @@
+	// is set to true in the delete options, the user must have permission
+	// to do 'delete-ignore-read-errors' on the given resource.
+	const verb = "delete-ignore-read-errors"
+	record := authorizer.AttributesRecord{
```
need teh field and label selectors.

tkashem replied:
yes, added the field and label selectors.

deads2k commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,141 @@
+	if decision == authorizer.DecisionAllow {
+		// TODO: no distinguishable attribute in the audit entry to
+		// discern whether it was a normal deletion flow or unsafe,
+		// add an annotation to audit to make that distinction?
```
Yes.  use the same annotation as the storage I'd think.

tkashem replied:
so for the admission, i have the following annotations that follow the admission convention:
```
allowunsafemalformedobjectdeletion.admission.k8s.io/decision: "allow|deny|empty when erroroccurs"
allowunsafemalformedobjectdeletion.admission.k8s.io/reason: "reason from the authorizer|internal error"
```

for the storage, I am using:
```
apiserver.k8s.io/unsafe-delete-ignore-read-error: ""
```

Let me know if you want to change it

deads2k commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,141 @@
+		// add an annotation to audit to make that distinction?
+		return nil
+	}
+
```
and audit here.

tkashem replied:
added

deads2k commented on pkg/registry/core/pod/storage/eviction_test.go:
```diff
@@ -581,30 +586,31 @@ func TestEvictionIgnorePDB(t *testing.T) {
 		},
 		{
-			name:            "with pdbs",
+			name:            "dry run - with pdbs",
```
why the name change here?

tkashem replied:
this test was originally named `TestEvictionDryRun` and each pre-existing test case was specific to dryRun

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -133,6 +136,13 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 		return nil, errors.NewBadRequest("name in URL does not match name in Eviction object")
 	}
 
+	if eviction.DeleteOptions != nil && ptr.Deref(eviction.DeleteOptions.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
```
is there not an validation.go validating evictions we can add thsi to?

tkashem replied:
That's where i wanted to add it, but I didn't find any validation here https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/policy/validation/validation.go, maybe because it is a sub resource of Pod?

tkashem commented:
/retest

```
{Failed;  === RUN   TestTerminationOrderingObeysGrace
    kuberuntime_termination_order_test.go:278: expected delay for container sc3 = 1, got 0
--- FAIL: TestTerminationOrderingObeysGrace (3.00s)
;}
```

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (d *corruptObjErrorInterpretingDecoder) Decode(value []byte, objPtr runtime.Object, rev int64) error {
+	// TODO: right now any error is deemed as undecodable, in
+	// the future, we can apply some filter, if need be.
```
yes, any error from here:
https://github.com/kubernetes/kubernetes/blob/daef8c2419a638d3925e146d0f5a6b217ea69b74/staging/src/k8s.io/apiserver/pkg/storage/etcd3/decoder.go#L61-L75

tkashem commented on plugin/pkg/admission/allowunsafedelete/admission.go:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	decision, reason, err := p.authz.Authorize(ctx, record)
+	if err != nil {
```
added `allowunsafemalformedobjectdeletion.admission.k8s.io/reason: "internal error"`

enj commented on plugin/pkg/admission/unsafedeletepolicy/admission.go:
```diff
@@ -0,0 +1,189 @@
+// PluginName is the name of the admission plugin
+const PluginName = "AllowUnsafeMalformedObjectDeletion"
+
+var (
+	prefix           = strings.ToLower(PluginName) + ".admission.k8s.io/"
+	auditDecisionKey = prefix + "decision"
+	auditReasonKey   = prefix + "reason"
+)
```
nit: just turn these into two `const`s.  It makes it easier to find via simple string searches.

enj reviewed (changes requested):
I just looked at the authz bits.

tkashem commented on plugin/pkg/admission/unsafedeletepolicy/admission.go:
```diff
@@ -0,0 +1,189 @@
+	if !found {
+		return admission.NewForbidden(attr, errors.New("no RequestInfo found in the context"))
+	}
+	if !requestInfo.IsResourceRequest || len(attr.GetSubresource()) > 0 {
+		return admission.NewForbidden(attr, errors.New("ignoreStoreReadErrorWithClusterBreakingPotential delete option is not allowed on a subresource or non-resource request"))
+	}
```
@deads2k any concern if we remove the admission plugin and do the access control inside the registry?

deads2k replied:
> Why is this an admission plugin instead of just part of the REST layer?

I'm not aware of the REST or storage layer taking an authorizer.  Perhaps it was done for SSA.  If there's precedent and it's clean, pushing it a layer deeper (and well gated) seems ok.

tkashem replied:
SSA patch does it here, i think
https://github.com/kubernetes/kubernetes/blob/50998de605d59bc02fe3ece4696d034e92a99e6a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L697-L704

I will look into moving it to the REST layer cleanly.

deads2k replied:
I found it here: https://github.com/kubernetes/kubernetes/blob/37ed906a33211c7d578cab2d681941ebfd2f2f23/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L232

yeah, pushing into endpoint/handlers is good.

tkashem replied:
this has been moved to endpoint/handlers https://github.com/kubernetes/kubernetes/pull/127513/commits/95179fb759e4cdb99ecf12793a7a2d93e73f0034

tkashem replied:
resolving it, please review the access control check inside endpoints/handlers

tkashem committed:
add access control check for unsafe delete

add access control check to ensure that the user has permission to do
'unsafe-delete-ignore-read-error' on the resource being deleted

tkashem commented on plugin/pkg/admission/unsafedeletepolicy/admission.go:
```diff
@@ -0,0 +1,189 @@
+		Resource:        attr.GetResource().Resource,
+		ResourceRequest: true,
+	}
+	// TODO: we should have a shared function to build this
```
done, thanks for the suggestion

tkashem replied:
it's not allowing me to use `ResourceAttributesFrom` due to `prevent staging --> k8s.io/kubernetes dep` rule in vendoring.

tkashem commented on plugin/pkg/admission/unsafedeletepolicy/admission.go:
```diff
@@ -0,0 +1,189 @@
+
+var _ admission.ValidationInterface = &UnsafeDeletePolicy{}
+var _ admission.InitializationValidator = &UnsafeDeletePolicy{}
+var _ genericadmissioninit.WantsAuthorizer = &UnsafeDeletePolicy{}
```
it's no longer an admission plugin, the access control has been moved to the registry layer

tkashem commented:
/retest

tkashem committed:
handle watch for unsafe delete

tkashem committed:
add an integration test that exercises the unsafe delete flow

- add an integration test that exercises the unsafe delete flow
- extend newTransformTest to enable RBAC
- add integration test to verify that LIST returns corrupt object keys

tkashem commented:
/retest

deads2k reviewed (changes requested):
I think I got about 3/4 done.  Some changes to gating requested.

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -122,12 +124,42 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
+
+				nextAdmission := deleteValiation
+				deleteValiation = func(ctx context.Context, old runtime.Object) error {
+					validator := unsafedeletepolicy.New(scope.Authorizer)
```
it's easier to write some tests for the admission logic if it's a separate unit. Not sure if I will have the time to refactor it before code freeze, I can definitely take care of it after the thaw.

deads2k replied:
> it's easier to write some tests for the admission logic if it's a separate unit. Not sure if I will have the time to refactor it before code freeze, I can definitely take care of it after the thaw.

It's easier to review and reason about inline.  The same concern exists for review time.

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -270,7 +303,7 @@ func (s *store) Delete(
 		return fmt.Errorf("unable to convert output object to pointer: %v", err)
 	}
 
-	skipTransformDecode := false
```
gated

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -912,8 +956,10 @@ func (s *store) getState(ctx context.Context, kv *mvccpb.KeyValue, key string, v
 		if skipTransformDecode {
-			// be explicit that we don't have the object
-			state.obj = nil
+			state.obj = reflect.New(v.Type()).Interface().(runtime.Object)
```
yes, gated

tkashem commented on staging/src/k8s.io/client-go/tools/cache/reflector_test.go:
```diff
@@ -1753,6 +1756,132 @@ func TestReflectorListExtract(t *testing.T) {
+	if want, got := 2, int(replaceInvoked.Load()); want != got {
+		t.Errorf("expected store Delete hooks to be invoked %d times, but got: %d", want, got)
+	}
+
```
yes, it has the correct data, and the test verifies that

tkashem commented on staging/src/k8s.io/client-go/tools/cache/reflector_test.go:
```diff
@@ -1753,6 +1756,132 @@ func TestReflectorListExtract(t *testing.T) {
+	}
+
+	lastExpectedRV := "3"
+	lists := []*v1.PodList{mkList("1"), mkList(lastExpectedRV)}
```
yes, added

tkashem commented:
/retest

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -122,12 +124,42 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
-		wasDeleted := true
 		userInfo, _ := request.UserFrom(ctx)
 		staticAdmissionAttrs := admission.NewAttributesRecord(nil, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Delete, options, dryrun.IsDryRun(options.DryRun), userInfo)
+		deleteValiation := rest.AdmissionToValidateObjectDeleteFunc(admit, staticAdmissionAttrs, scope)
```
fixed

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector_test.go:
```diff
@@ -1753,6 +1756,152 @@ func TestReflectorListExtract(t *testing.T) {
 	}
 }
 
+func TestReflectorWithUnsafeDelete(t *testing.T) {
```
what is it testing? The name should better be a claim, not a feature description.

sttts replied:
`TestReflectorRecoversWithRightRVOnUnsafeDelete` or something like that.

sttts commented on test/integration/controlplane/transformation/secrets_transformation_test.go:
```diff
@@ -99,6 +120,513 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+			// i) wait for the breaking changes to take effect
+			testCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			err = wait.PollUntilContextTimeout(testCtx, 1*time.Second, 2*time.Minute, true, func(ctx context.Context) (done bool, err error) {
```
```suggestion
			err = wait.PollUntilContextTimeout(testCtx, 1*time.Second, wait.ForeverTestTimeout, true, func(ctx context.Context) (done bool, err error) {
```

tkashem replied:
unfortuately dynamic encryption config reload takes about 1m, so can't use wait.ForeverTestTimeout just yet, I have left to TODO to investigate and improve the reload time
```

deads2k replied:
Add it to beta criteria or we will lose it.

sttts commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -133,6 +136,13 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 
+	if eviction.DeleteOptions != nil && ptr.Deref(eviction.DeleteOptions.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
+		errs := field.ErrorList{
+			field.Invalid(field.NewPath("deleteOptions").Child("ignoreStoreReadErrorWithClusterBreakingPotential"), true, "can not be set for pod eviction, try after removing it"),
```
removing what?

tkashem replied:
fixed the message

tkashem commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go:
```diff
@@ -0,0 +1,141 @@
+
+	// let's make sure that the audit will reflect that this delete request
+	// was tried with ignoreStoreReadErrorWithClusterBreakingPotential enabled
+	audit.AddAuditAnnotation(ctx, "apiserver.k8s.io/unsafe-delete-ignore-read-error", "")
```
if i am not mistaken, admission/access check will happen at the storage layer:
https://github.com/kubernetes/kubernetes/blob/33c64b380a11ba5a4d18a55b73d5f0a3ce89dc85/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L323

so I think we should see this annotation even when authz fails

deads2k replied:
Please move this annotation to the endpoint handler.

tkashem replied:
moved

tkashem commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -358,3 +361,31 @@ func isValidConditionReason(value string) []string {
+
+// ValidateIgnoreStoreReadError validates that delete options are valid when
+// ignoreStoreReadErrorWithClusterBreakingPotential is enabled
+func ValidateIgnoreStoreReadError(fldPath *field.Path, options *metav1.DeleteOptions) field.ErrorList {
```
nothing at the moment, it will have no effect, do we want me to add a validation inside `DeleteCollection` handler?

deads2k replied:
> nothing at the moment, it will have no effect, do we want me to add a validation inside `DeleteCollection` handler?

We should ensure the value isn't set for DeleteCollection calls.

tkashem replied:
yes, validation added

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -464,7 +464,17 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
+	store := etcd3.New(client, c.Codec, newFunc, newListFunc, c.Prefix, resourcePrefix, c.GroupResource, transformer, c.LeaseManagerConfig, decoder, versioner)
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion) {
+		store = etcd3.NewStoreWithUnsafeCorruptObjectDeletion(store, c.GroupResource)
+	}
```
moved

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -464,7 +464,17 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
+		transformer = etcd3.WithCorruptObjErrorHandlingTransformer(transformer)
+		decoder = etcd3.WithCorruptObjErrorHandlingDecoder(decoder)
+	}
+
```
moved

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -914,6 +966,12 @@ func (s *store) getState(ctx context.Context, kv *mvccpb.KeyValue, key string, v
 			// be explicit that we don't have the object
 			state.obj = nil
+			if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
+				state.obj = reflect.New(v.Type()).Interface().(runtime.Object)
```
this precondition check commit has been removed

tkashem commented:
/retest

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -122,12 +128,40 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
+				nextAdmission := deleteValidation
+				deleteValidation = func(ctx context.Context, old runtime.Object) error {
+					// we first check if the user has permission to do unsafe delete
+					if err := authorizeUnsafeDelete(ctx, staticAdmissionAttrs, scope.Authorizer); err != nil {
```
Let's move this out of the `deleteValidation` method and instead do it on line 147

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -294,3 +328,100 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope *RequestSc
+	}
+
+	const (
+		auditDecisionKey = "allowunsafemalformedobjectdeletion.admission.k8s.io/decision"
+		auditReasonKey   = "allowunsafemalformedobjectdeletion.admission.k8s.io/reason"
```
I don't think we need this level of auditing.  If the request is rejected or allowed, we'll see the status code in the log. The big need was, "was this delete an unsafe one".

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -122,12 +128,40 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
 		staticAdmissionAttrs := admission.NewAttributesRecord(nil, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Delete, options, dryrun.IsDryRun(options.DryRun), userInfo)
+		deleteValidation := rest.AdmissionToValidateObjectDeleteFunc(admit, staticAdmissionAttrs, scope)
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
```
nearly forgot.  As soon as we decode the delete options above, if the feature is disabled, nil the field.

tkashem commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -358,3 +361,31 @@ func isValidConditionReason(value string) []string {
+		return allErrs
+	}
+
+	if len(options.DryRun) > 0 {
```
this is a todo for post thaw/beta, we also want to provide error message `IgnoreStoreReadErrorWithClusterBreakingPotential can not be used together with {dryRun|gracePerioSeconds}`

deads2k replied:
> this is a todo for post thaw/beta, we also want to provide error message `IgnoreStoreReadErrorWithClusterBreakingPotential can not be used together with {dryRun|gracePerioSeconds}`

Link to where it's added to beta criteria please.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -265,6 +301,16 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope *RequestSc
 			return
 		}
+
+		if options != nil && ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, true) {
```
needs to be featuregated.  Also after we have deleteoptions, if the gate is off, need to set the `IgnoreStoreReadErrorWithClusterBreakingPotential` to nil.

deads2k commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -54,6 +54,15 @@ const (
 	// Allows us to enable anonymous auth for only certain apiserver endpoints.
 	AnonymousAuthConfigurableEndpoints featuregate.Feature = "AnonymousAuthConfigurableEndpoints"
 
+	// owner: @stlaz @tkashem @dgrisonnet
```
sounded like standa wanted to be removed.

tkashem replied:
no, he is fine to be here, we talked to him

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -295,3 +341,77 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope *RequestSc
+// 'unsafe-delete-ignore-read-errors' on the resource being deleted when
+// ignoreStoreReadErrorWithClusterBreakingPotential is enabled
+func authorizeUnsafeDelete(ctx context.Context, attr admission.Attributes, authz authorizer.Authorizer) (err error) {
+	if attr.GetOperation() != admission.Delete || attr.GetOperationOptions() == nil {
+		return nil
+	}
+	options, ok := attr.GetOperationOptions().(*metav1.DeleteOptions)
+	if !ok {
+		return errors.NewInternalError(fmt.Errorf("expected an option of type: %T, but got: %T", &metav1.DeleteOptions{}, attr.GetOperationOptions()))
+	}
+	if !ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
+		return nil
+	}
```
these are all always unnecessary given it's only called from the delete resthandler, right?

deads2k reviewed (changes requested):
need gating fixes.

Also need links to where beta criteria have added the validation I called out here: https://github.com/kubernetes/kubernetes/pull/127513#discussion_r1832826581

deads2k commented:
> Also need links to where beta criteria have added the validation I called out here: [#127513 (comment)](https://github.com/kubernetes/kubernetes/pull/127513#discussion_r1832826581)

https://github.com/kubernetes/enhancements/pull/4949/files

deads2k commented:
/lgtm
/approve

deads2k commented:
/milestone v1.32

tkashem commented:
repushed,  had to run `make update`  cc @dgrisonnet

dgrisonnet commented:
/lgtm

liggitt commented:
Apparently this is failing e2e alpha and beta jobs, see https://github.com/kubernetes/kubernetes/issues/128704

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -260,11 +296,26 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope *RequestSc
 		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
+			if options != nil && ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, true) {
```
This invocation of ptr.Deref is incorrect, this is defaulting to true if nil and is breaking tests.

liggitt replied:
This is causing https://github.com/kubernetes/kubernetes/issues/128704#issuecomment-2466004961
